input/taproot_test.go

```diff
@@ -0,0 +1,1871 @@
+			HashType:          sigHash,
+			InputIndex:        0,
+			SigHashes:         hashCache,
+			SignMethod:        TaprootScriptSpendSignMethod,
+			PrevOutputFetcher: prevOuts,
+		}
+
+		return SenderHTLCScriptTaprootTimeout(
+			receiverSig, sigHash, signer, signDesc, spendTx,
+			htlcScriptTree.revokeKey.PubKey(),
+			scriptTree.TapscriptTree,
+		)
+	}
+}
+
+// TestTaprootSenderHtlcSpend tests that all the positive and negative paths
+// for the sender HTLC tapscript tree work as expected.
+func TestTaprootSenderHtlcSpend(t *testing.T) {
+	t.Parallel()
+
+	// First, create a new test script tree.
+	htlcScriptTree := newTestSenderHtlcScriptTree(t)
+
+	spendTx := wire.NewMsgTx(2)
+	spendTx.AddTxIn(&wire.TxIn{})
+	spendTx.AddTxOut(&wire.TxOut{
+		Value: htlcScriptTree.htlcAmt,
+	})
+
+	testCases := []struct {
+		name string
+
+		witnessGen witnessGen
+
+		txInMutator func(txIn *wire.TxIn)
+
+		witnessMutator func(witness wire.TxWitness)
+
+		valid bool
+	}{
+		// Valid redeem with the pre-image, and the spending
+		// transaction set to CSV 1 to enforce the required delay.
+		{
+			name: "redeem success valid sighash all",
+			witnessGen: htlcSenderRedeemValidWitnessGen(
+				txscript.SigHashAll, htlcScriptTree,
+			),
+			txInMutator: func(txIn *wire.TxIn) {
+				txIn.Sequence = 1
+			},
+			valid: true,
+		},
+
+		// Valid with pre-image, using sighash default.
+		{
+			name: "redeem success valid sighash default",
+			witnessGen: htlcSenderRedeemValidWitnessGen(
+				txscript.SigHashDefault, htlcScriptTree,
+			),
+			txInMutator: func(txIn *wire.TxIn) {
+				txIn.Sequence = 1
+			},
+			valid: true,
+		},
+
+		// Valid with pre-image, using sighash single+anyonecanpay.
+		{
+			name: "redeem success valid sighash " +
+				"single|anyonecanpay",
+			witnessGen: htlcSenderRedeemValidWitnessGen(
+				txscript.SigHashSingle|
+					txscript.SigHashAnyOneCanPay,
+				htlcScriptTree,
+			),
+			txInMutator: func(txIn *wire.TxIn) {
+				txIn.Sequence = 1
+			},
+			valid: true,
+		},
+
+		// Invalid spend, the witness is correct, but the spending tx
+		// doesn't have a sequence of 1 set. This uses the CSV 0 trick:
+		// 0 > 0 -> false.
+		{
+			name: "redeem success invalid wrong sequence",
+			witnessGen: htlcSenderRedeemValidWitnessGen(
+				txscript.SigHashAll, htlcScriptTree,
+			),
+			valid: false,
+		},
+
+		// Valid spend with the revocation key, sighash all.
+		{
+			name: "revocation spend vaild sighash all",
+			witnessGen: htlcSenderRevocationWitnessGen(
+				txscript.SigHashAll, htlcScriptTree,
+			),
+			valid: true,
+		},
+
+		// Valid spend with the revocation key, sighash default.
+		{
+			name: "revocation spend vaild sighash default",
+			witnessGen: htlcSenderRevocationWitnessGen(
+				txscript.SigHashDefault, htlcScriptTree,
+			),
+			valid: true,
+		},
+
+		// Invalid spend with the revocation key. The witness mutator
+		// modifies the sig.
+		{
+			name: "revocation spend invalid",
```

Comment:
add a `sighash single` case for revocation path?

Reply:
Done!